File: solana_toolbox_idl/src/toolbox_idl_service.rs
use std::collections::HashMap;
use std::sync::Arc;

use anyhow::Context;
use anyhow::Result;
use solana_sdk::pubkey::Pubkey;
use solana_toolbox_endpoint::ToolboxEndpoint;

use crate::toolbox_idl_program::ToolboxIdlProgram;

pub struct ToolboxIdlService {
    cached_programs: HashMap<Pubkey, Option<Arc<ToolboxIdlProgram>>>,
}

impl Default for ToolboxIdlService {
    fn default() -> ToolboxIdlService {
        ToolboxIdlService::new()
    }
}

impl ToolboxIdlService {
    pub fn new() -> ToolboxIdlService {
        ToolboxIdlService {
            cached_programs: Default::default(),
        }
    }

    pub fn set_program(
        &mut self,
        program_id: &Pubkey,
        idl_program: Option<Arc<ToolboxIdlProgram>>,
    ) {
        self.cached_programs.insert(*program_id, idl_program);
    }

    pub async fn get_or_resolve_program(
        &mut self,
        endpoint: &mut ToolboxEndpoint,
        program_id: &Pubkey,
    ) -> Result<Option<Arc<ToolboxIdlProgram>>> {
        if let Some(idl_program) = self.cached_programs.get(program_id) {
            return Ok(idl_program.clone());
        }
        let idl_program =
            ToolboxIdlService::resolve_program(endpoint, program_id).await?;
        self.cached_programs
            .insert(*program_id, idl_program.clone());
        Ok(idl_program)
    }

    async fn resolve_program(
        endpoint: &mut ToolboxEndpoint,
        program_id: &Pubkey,
    ) -> Result<Option<Arc<ToolboxIdlProgram>>> {
        if let Some(idl_program) = ToolboxIdlProgram::from_lib(program_id) {
            return Ok(Some(Arc::new(idl_program)));
        }
        Ok(endpoint
            .get_account(
                &ToolboxIdlProgram::find_anchor_address(program_id)
                    .context("Find Anchor Idl Account")?,
            )
            .await
            .context("Get Anchor Idl Account")?
            .map(|source_account| {
                ToolboxIdlProgram::try_parse_from_account_data(
                    &source_account.data,
                )
            })
            .transpose()
            .context("Parse IDL Account Data")?
            .map(Arc::new))
    }
}

File: solana_toolbox_web3/src/ToolboxIdlService.ts
import {
  AccountInfo,
  PublicKey,
  SystemProgram,
  TransactionInstruction,
} from '@solana/web3.js';
import { ToolboxIdlProgram } from './ToolboxIdlProgram';
import { ToolboxEndpoint } from './ToolboxEndpoint';
import { ToolboxIdlAccount } from './ToolboxIdlAccount';
import { ToolboxIdlInstruction } from './ToolboxIdlInstruction';

export class ToolboxIdlService {
  private cachedPrograms: Map<PublicKey, ToolboxIdlProgram | null>;

  constructor() {
    this.cachedPrograms = new Map<PublicKey, ToolboxIdlProgram | null>();
  }

  public setProgram(
    programId: PublicKey,
    idlProgram: ToolboxIdlProgram | null,
  ) {
    this.cachedPrograms.set(programId, idlProgram);
  }

  public async getOrResolveProgram(
    endpoint: ToolboxEndpoint,
    programId: PublicKey,
  ): Promise<ToolboxIdlProgram | null> {
    let cachedProgram = this.cachedPrograms.get(programId);
    if (cachedProgram !== undefined) {
      return cachedProgram;
    }
    let resolvedProgram = await ToolboxIdlService.resolveProgram(
      endpoint,
      programId,
    );
    this.cachedPrograms.set(programId, resolvedProgram);
    return resolvedProgram;
  }

  static async resolveProgram(
    endpoint: ToolboxEndpoint,
    programId: PublicKey,
  ): Promise<ToolboxIdlProgram | null> {
    // TODO - lib idls
    let account = await endpoint.getAccount(
      await ToolboxIdlProgram.findAnchorAddress(programId),
    );
    if (account == null) {
      return null;
    }
    return ToolboxIdlProgram.tryParseFromAccountData(account.data);
  }

  public async getAndDecodeAccount(
    endpoint: ToolboxEndpoint,
    address: PublicKey,
  ) {
    let account = (await endpoint.getAccount(address)) ?? {
      lamports: 0,
      owner: SystemProgram.programId,
      data: Buffer.from([]),
      executable: false,
    };
    return this.decodeAccount(endpoint, account);
  }

  public async decodeAccount(
    endpoint: ToolboxEndpoint,
    account: AccountInfo<Buffer>,
  ) {
    let idlProgram =
      (await this.getOrResolveProgram(endpoint, account.owner)) ??
      ToolboxIdlProgram.Unknown;
    let idlAccount =
      idlProgram.guessAccount(account.data) ?? ToolboxIdlAccount.Unknown;
    let accountState = idlAccount.decode(account.data);
    return {
      lamports: account.lamports,
      owner: account.owner,
      program: idlProgram,
      account: idlAccount,
      state: accountState,
    };
  }

  public async decodeInstruction(
    endpoint: ToolboxEndpoint,
    instruction: TransactionInstruction,
  ) {
    let idlProgram =
      (await this.getOrResolveProgram(endpoint, instruction.programId)) ??
      ToolboxIdlProgram.Unknown;
    let idlInstruction =
      idlProgram.guessInstruction(instruction.data) ??
      ToolboxIdlInstruction.Unknown;
    let { instructionProgramId, instructionAddresses, instructionPayload } =
      idlInstruction.decode(instruction);
    return {
      program: idlProgram,
      instruction: idlInstruction,
      instructionProgramId,
      instructionAddresses,
      instructionPayload,
    };
  }

  // TODO - support finding
}
